src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,156 @@
+        returnNode(index).value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeDeleted = returnNode(index);
+        unlink(nodeDeleted);
+        return nodeDeleted.value;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            Node<T> nodeDeleted = getNode(i);
+            T nodeValue = nodeDeleted.value;
+            if (nodeValue != null && nodeValue.equals(object) || nodeValue == object) {
+                unlink(nodeDeleted);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndex(index);
+        Node<T> pointer;
+        if (index < size / 2) {
+            pointer = head;
+            for (int i = 0; i < size / 2; i++) {
+                if (i == index) {
+                    return pointer;
+                }
+                pointer = pointer.next;
+            }
+        } else {
+            pointer = tail;
+            for (int i = size - 1; i >= size / 2; i--) {
+                if (i == index) {
+                    return pointer;
+                }
+                pointer = pointer.prev;
+            }
+        }
+        return pointer;
+    }
+
+    private void linkBefore(T value, Node<T> target) {
+        if (target == null) {
+            throw new NullPointerException("Node cannot be null!");
+        }
+        Node<T> nodeToLink = new Node<>(target.prev, value, target);
+        if (target.prev == null) {
+            head = nodeToLink;
+        } else {
+            target.prev.next = nodeToLink;
+        }
+        target.prev = nodeToLink;
+        size++;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index
+                    + "passed to the method is invalid. Actual size is " + size);
+        }
+    }
+
+    private void unlink(Node<T> nodeDeleted) {
+        if (nodeDeleted == head) {
+            head = nodeDeleted.next;
+        } else if (nodeDeleted == tail) {
+            tail = nodeDeleted.prev;
+        } else {
+            nodeDeleted.prev.next = nodeDeleted.next;
+            nodeDeleted.next.prev = nodeDeleted.prev;
+        }
+        size--;
+    }
+
+    private Node<T> returnNode(int index) {
+        checkIndex(index);
+        return getNode(index);
```

Comment:
```suggestion
```
Why did you add it???

Reply:
have no idea why :)
Hope I added it before first changes. And just haven't payed attention that i added checkIndex in getNode method.